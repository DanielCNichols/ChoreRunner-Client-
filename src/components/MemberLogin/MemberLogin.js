import React, { Component } from 'react'
import AuthApiService from '../../services/auth-api-service'
import UserContext from '../../contexts/UserContext'
import './MemberLogin.css'

export default class MemberLogin extends Component {

  static contextType = UserContext

  state = {
    username: '',
    password: '',
    error: null,
    validateError: {
      usernameError: '',
      passwordError:'',
    }
  }

  onChangeHandle = (e) => {
    this.setState({
      [e.target.name]: e.target.value
    })
  }

  validate = () => {
    const REGEX_UPPER_LOWER_NUMBER_SPECIAL = /(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[!@#\$%\^&])[\S]+/;
    const { username, password } = this.state;
    let usernameError = '';
    let passwordError = '';
  
    if(username.trim().length <= 5) {
      usernameError = 'Please enter more characters'
    }
    
    if (password.length < 5) {
      passwordError = 'Password be longer than 8 characters';
    }
    if (password.length > 20) {
      passwordError = 'Password be less than 72 characters';
    }
    if (password.startsWith(' ') || password.endsWith(' ')) {
      passwordError = 'Password must not start or end with empty spaces';
    }
    if (!REGEX_UPPER_LOWER_NUMBER_SPECIAL.test(password)) {
      passwordError = 'Password must contain one upper case, lower case, number and special character';
    }


    if(passwordError || usernameError) {
      this.setState({ validateError: { passwordError, usernameError } })
      return false;
    }
    return true;
  }

  onLoginSuccess = () => {
    const { location, history } = this.props
    const destination = (location.state || {}).from || '/member-dashboard'
    history.push(destination)
  }
  

  handleSubmit = (e) => {
    e.preventDefault()
    const { username, password } = e.target
    const isValid = this.validate();

    this.setState({ error: null })

   // if(isValid) {
      AuthApiService.postMemberLogin({
        username: username.value,
        password: password.value,
      })
      .then(res => {
        username.value = ''
        password.value = ''
        this.context.processLogin(res.authToken, res.type)
        this.onLoginSuccess()
      })
      .catch(res => {
        this.setState({ error: res.error })
      })
      document.getElementById("member-login").reset();
   // }

  }
  render() {
    const { error, username, password } = this.state
    const { usernameError, passwordError } = this.state.validateError

    return (
      <div className='parent-login'>
        <h2> Member login</h2>
        <form className='parent-form-container' id="member-login" onSubmit={this.handleSubmit}>
          {/* <div role='alert'>
            {error && <p>{error}</p>}
          </div> */}
          <div>
            <label htmlFor='username'> Username</label>
            <input name='username' type='text' onChange={this.onChangeHandle} value={username} required ></input>
            <div className="valid-error">{usernameError}</div>
          </div>
          <div>
            <label htmlFor='password'> Password</label>
            <input name='password' type='password' onChange={this.onChangeHandle} value={password} required ></input>
            <div role='alert'>
              {error && <p>{error}</p>}
            </div>
          </div>
          <button type='submit'>login</button>
        </form>
      </div>
    )
  }
}

