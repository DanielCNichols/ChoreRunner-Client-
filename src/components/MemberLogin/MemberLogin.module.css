@import '../../vars.css';

.error {
  color: red;
  margin-bottom: 5px;
}
.container {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  height: 95vh;
  margin: 0 auto 50px auto;
  padding-top: 100px;
}

.memberForm {
  border: 2px solid white;
  background-color: var(--color-menus);
  border-radius: 6px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 90%;
  padding: 25px 0;
}

.memberForm fieldset {
  display: flex;
  flex-direction: column;
  justify-content: center;
  width: 70%;
}

.memberForm legend {
  font-family: var(--font-headers);
  color: var(--color-mainHeader);
  margin-bottom: 20px;
  font-size: 1.3em;
}

.memberForm label {
  margin-bottom: 5px;
  font-family: var(--font-body);
  color: var(--color-bodyText);
}

.formElement {
  margin-bottom: 20px;
}

div[role='alert'] > p {
  color: white;
  width: auto;
  margin: auto;
  font-size: 1.1rem;
  font-weight: 500;
  font-family: 'Electrolize', sans-serif;
}

.memberForm button {
  width: 120px;
  text-align: center;
  margin: 20px auto 0 auto;
  font-family: 'Press Start 2P', cursive;
  font-size: 20px;
  padding: 7px;
  background-color: rgb(255, 217, 0);
  text-shadow: -1px -1px rgb(105, 90, 0), 1px 1px white;
  color: #725f00;
  -webkit-border-radius: 7px;
  -moz-border-radius: 7px;
  -o-border-radius: 7px;
  border-radius: 7px;
  box-shadow: 0 0.3em rgb(134, 112, 0);
  cursor: pointer;
}

.demo {
  width: 260px;
  margin: 0 auto 30px auto;
  display: flex;
  font-family: 'Electrolize', sans-serif;
  color: white;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  padding: 0px 10px 30px 10px;
  background: rgb(0, 0, 0, 0.5);
  position: relative;
  animation: bounce 500ms infinite alternate;
  -webkit-animation: bounce 500ms infinite alternate;
  border-radius: 10px;
}
.demo p {
  margin: 0;
}

@media (min-width: 450px) {
  .memberForm {
    width: 40%;
    min-width: 350px;
  }
}

@media (min-width: 875px) {
  .memberForm {
    width: 400px;
  }
}
