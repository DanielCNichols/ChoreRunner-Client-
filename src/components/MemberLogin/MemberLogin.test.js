import React from 'react';
import ReactDOM from 'react-dom';
import MemberLogin from './MemberLogin';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';

describe('MemberLogin component testing', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');

    ReactDOM.render(
      <MemoryRouter>
        <MemberLogin />
      </MemoryRouter>,
      div
    );

    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders UI as expected', () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <MemberLogin />
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
