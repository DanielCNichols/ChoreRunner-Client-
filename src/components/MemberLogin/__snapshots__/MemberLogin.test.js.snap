// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MemberLogin component testing renders UI as expected 1`] = `
<div
  className="container"
>
  <form
    className="memberForm"
    onSubmit={[Function]}
  >
    <div
      className="demo"
    >
      <h3>
        Demo:
      </h3>
      <p>
        username: margeincharge
      </p>
      <p>
        password: Password123!
      </p>
    </div>
    <fieldset
      className="Fieldset"
    >
      <legend
        className="Legend"
      >
        Member Login
      </legend>
      <div
        className="FormElement formElement"
      >
        <label
          className="Label"
          htmlFor="username"
        >
          Username
        </label>
        <input
          className="Input"
          name="username"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <div
        className="FormElement formElement"
      >
        <label
          htmlFor="password"
        >
          Password
        </label>
        <input
          className="Input"
          name="password"
          onChange={[Function]}
          type="password"
          value=""
        />
      </div>
      <button
        className="arcadeButton"
        type="submit"
      >
        login
      </button>
    </fieldset>
  </form>
</div>
`;
